File: data structures and algorithms/cpp/04_016.cpp
#include <memory>

class Coffee {
public:
    virtual ~Coffee() = default;
    virtual double getCost() = 0;
};

class SimpleCoffee : public Coffee {
public:
    double getCost() override {
        return 1.0;
    }
};

class MilkDecorator : public Coffee {
private:
    std::unique_ptr<Coffee> coffee;
    
public:
    MilkDecorator(std::unique_ptr<Coffee> c) 
        : coffee(std::move(c)) {}
    
    double getCost() override {
        return coffee->getCost() + 0.5;
    }
};

File: data structures and algorithms/cs/04_016.cs
public interface ICoffee
{
    double GetCost();
}

public class SimpleCoffee : ICoffee
{
    public double GetCost() { return 1.0; }
}

public class MilkDecorator : ICoffee
{
    private ICoffee coffee;

    public MilkDecorator(ICoffee coffee)
    {
        this.coffee = coffee;
    }

    public double GetCost()
    {
        return coffee.GetCost() + 0.5;
    }
}
